src/repositories/PriceRepository.ts

```diff
@@ -1,41 +1,120 @@
+  cache: NodeCache;
+  defaultPriceTTL: number;
+
+  constructor(@inject('Settings') settings: Settings) {
+    this.cache = new NodeCache({stdTTL: 60, checkperiod: 60});
+    this.defaultPriceTTL = settings.mongodb.indexTTL.priceTTL;
+  }
+
+  async saveBatch(props: SavePriceProps[]): Promise<void> {
+    const operations = [];
+
+    for (const priceProp of props) {
+      const {source, symbol, value, timestamp} = priceProp;
+      const expireAt = priceProp.expireAt || dayjs().add(this.defaultPriceTTL, 'second').toDate();
+
+      operations.push({
+        updateOne: {
+          filter: {source, symbol, timestamp},
+          update: {source, symbol, timestamp, value, expireAt},
+          upsert: true,
+          new: true,
+        },
+      });
+    }
+
+    await getModelForClass(Price).bulkWrite(operations);
+  }
+
+  async getLatestPrice(props: LatestPriceProps): Promise<number | undefined> {
+    return (await this.getLatestPriceRecord(props))?.value;
+  }
+
+  async getLatestPriceRecord(props: LatestPriceProps): Promise<Price | undefined> {
+    const {
+      source,
+      symbol,
+      timestamp: {from, to},
+    } = props;
+
+    const timestampFilter: {[key: string]: Date} = {};
+    if (from) timestampFilter['$gte'] = from;
+    if (to) timestampFilter['$lt'] = to;
 
-  constructor(@inject(PriceAggregator) priceAggregator: PriceAggregator) {
-    this.priceAggregator = priceAggregator;
+    const price = await getModelForClass(Price)
+      .findOne(
+        {
+          source,
+          symbol,
+          timestamp: timestampFilter,
+        },
+        {},
+        {sort: {timestamp: -1}},
+      )
+      .exec();
+
+    return price || undefined;
   }
 
-  async savePrice(prefix: string, pair: string, price: number, timestamp: number): Promise<void> {
-    const formattedPair = pair.toUpperCase().replace('-', '~');
+  async getValueAndTimestamp(props: LatestPriceProps): Promise<{value: number; timestamp: number} | null> {
+    const {source, symbol, timestamp} = props;
+    const price = await this.getLatestPriceRecord({
+      source,
+      symbol,
+      timestamp,
+    });
 
-    return this.priceAggregator.add(`${prefix}${formattedPair}`, price, timestamp);
+    return price ? {timestamp: Math.floor(price.timestamp.getTime() / 1000), value: price.value} : null;
   }
 
-  async getLatestPrice(
-    prefix: string,
-    {fsym, tsym, freshness = this.settings.api.priceFreshness}: PairWithFreshness,
-    timestamp: number,
-  ): Promise<number | null> {
-    return this.priceAggregator.valueAfter(`${prefix}${fsym}~${tsym}`, timestamp, timestamp - freshness);
+  async getValueTimestamps(symbol: string, source: string): Promise<{value: number; timestamp: number}[]> {
+    const prices = await getModelForClass(Price).find({source, symbol}).exec();
+    return prices.map((price) => ({timestamp: Math.floor(price.timestamp.getTime() / 1000), value: price.value}));
   }
 
   async getLatestPrices(
```

Comment:
looks like `getLatestPrices` method is not in use

Reply:
Removed